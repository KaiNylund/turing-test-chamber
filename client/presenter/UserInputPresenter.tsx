import { Component } from "react";
import GuessButtonView from "../view/GuessButtonView";
import GuesserInputBarView from "../view/GuesserInputBarView";
import { DefaultStateType } from "../types";
import { connect } from "react-redux";
import store from "../model/store";
import React from "react";
import { inputTextChange, submitMessage, startGuessing } from "../model/actions";

const mapStateToProps = (state: DefaultStateType) => {
  return {
    inputText: state.inputText,
    isGuessing: state.isGuessing,
    waitingForPlayerMessage: state.waitingFor.playerMessage,
    waitingForGuessOutcome: state.waitingFor.guessOutcome,
    guessOutcome: state.guessOutcome
  }
}

const mapDispatchToProps = {
  submitMessage,
  inputTextChange,
  startGuessing
}

class UserInputPresenter extends Component<{inputText: string,
                                            isGuessing: boolean,
                                            waitingForPlayerMessage: boolean,
                                            waitingForGuessOutcome: boolean,
                                            guessOutcome: "undecided" | "correct" | "incorrect",
                                            submitMessage: (e: any) => any,
                                            inputTextChange: (e: any) => any,
                                            startGuessing: (e: any) => any}, {}> {
  render() {
    return (
      <div className={"guessing-inputs" + ((this.props.isGuessing) ? " is-guessing" : "")}>
        <GuesserInputBarView onGuesserTextChange={this.props.inputTextChange}
                             onGuesserSubmit={this.props.submitMessage}
                             inputText={this.props.inputText}
                             disabled={this.props.isGuessing ||
                                       this.props.waitingForPlayerMessage ||
                                       this.props.waitingForGuessOutcome ||
                                       this.props.guessOutcome !== "undecided"}/>
        {store.getState().role.type === "guesser" &&
        <GuessButtonView startGuessing={this.props.startGuessing}
                         disabled={this.props.isGuessing ||
                                   this.props.guessOutcome !== "undecided"}/>}
      </div>
    )
  }

}

export default connect(mapStateToProps, mapDispatchToProps)(UserInputPresenter);