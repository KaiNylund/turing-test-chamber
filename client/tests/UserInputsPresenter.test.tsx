import React from 'react';
import renderer from 'react-test-renderer';
import UserInputPresenter from '../presenter/UserInputPresenter';
import { Provider } from 'react-redux';
import store from '../model/store';

// NOTE: need to recreate component in each call becasue we're updating the store each time
// (e.g. disabling after a submit should not affect the next time we test submitting)

// Test input
it('input text updates', async () => {
  const component = renderer.create(
    <Provider store={store}>
      <UserInputPresenter/>
    </Provider>
  );
  let tree: any = component.toJSON();
  let formComponent = tree.children[0];
  let inputComponent = formComponent.children.find((e: { type: string; }) => e.type === "input");
  expect(tree).toMatchSnapshot();

  // NOTE: There's a bug in Formik that throws a warning saying we aren't using .act() unless
  // we wrap our expect calls in waitFor.
  await renderer.act(async () => {
    inputComponent.props.onChange({target: { value: "testing again..."}});
  });

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});


// Test submit updates
it('input and button disabled on submit', async () => {
  const component = renderer.create(
    <Provider store={store}>
      <UserInputPresenter/>
    </Provider>
  );

  let tree: any = component.toJSON();
  let formComponent = tree.children[0];
  let inputComponent = formComponent.children.find((e: { type: string; }) => e.type === "input");
  expect(tree).toMatchSnapshot();

  // NOTE: There's a bug in Formik that throws a warning saying we aren't using .act() unless
  // we wrap our expect calls in waitFor.
  await renderer.act(async () => {
    inputComponent.props.onChange({target: { value: "testing..."}});
  });

  // General submit
  await renderer.act(async () => {
    formComponent.props.onSubmit();
  });

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});


// Test submit button
it('submit button works', async () => {
  const component = renderer.create(
    <Provider store={store}>
      <UserInputPresenter/>
    </Provider>
  );

  let tree: any = component.toJSON();
  let formComponent = tree.children[0];
  let inputComponent = formComponent.children.find((e: { type: string; }) => e.type === "input");
  let submitButton = formComponent.children.find((e: { type: string; }) => e.type === "button");
  expect(tree).toMatchSnapshot();

  await renderer.act(async () => {
    inputComponent.props.onChange({target: { value: "testing..."}});
  });

  // Submit by clicking on the "submit button"
  await renderer.act(async () => {
    submitButton.props.onClick();
  });

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});