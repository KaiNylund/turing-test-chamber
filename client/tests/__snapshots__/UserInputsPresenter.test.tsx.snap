// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input and button disabled on submit 1`] = `
<div
  className="guessing-inputs"
>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <input
      disabled={false}
      name="chat-input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Type here..."
      tabIndex={-1}
      type="text"
      value="testing again..."
    />
    <button
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`input and button disabled on submit 2`] = `
<div
  className="guessing-inputs"
>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <input
      disabled={true}
      name="chat-input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Type here..."
      tabIndex={-1}
      type="text"
      value=""
    />
    <button
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`input text updates 1`] = `
<div
  className="guessing-inputs"
>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <input
      disabled={false}
      name="chat-input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Type here..."
      tabIndex={-1}
      type="text"
      value=""
    />
    <button
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`input text updates 2`] = `
<div
  className="guessing-inputs"
>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <input
      disabled={false}
      name="chat-input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Type here..."
      tabIndex={-1}
      type="text"
      value="testing again..."
    />
    <button
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`submit button works 1`] = `
<div
  className="guessing-inputs"
>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <input
      disabled={true}
      name="chat-input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Type here..."
      tabIndex={-1}
      type="text"
      value=""
    />
    <button
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`submit button works 2`] = `
<div
  className="guessing-inputs"
>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <input
      disabled={true}
      name="chat-input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Type here..."
      tabIndex={-1}
      type="text"
      value="testing..."
    />
    <button
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
